## Prior Feedback
### Issue 1.1
The dataframe produced by alignSNPs.py has the following information:
- variantChromosome (Chromosome the variant is located on), variantPosition (Base pair location of the variant), variantAlt (The alternative base found for a given variant), ensemblID (The gene ID a variant is located within), variantPIP (The probability that a particular variant is causal for a signal within a given credible set), NMATCH (If the variant was found to have introgressed from Neanderthals), DMATCH (If the variant was found to have introgressed from Denisovans), and POPGROUP (The population group in which a given introgression occured)

The log file for alignSNPs.py returns the number of matching SNPs found within a set search space in the following format:
- **Number of Matches Found** Alignments were found out of **Average number of introgressions of the searched-for match type across all population groups** introgressions across **Number of SNPs surveyed from selected eQTL finemapping dataset** SNPs.

Together, this information can be used in downstream analysis to filter for any selection of fine-mapped SNPs from any population group.

### Issue 1.2
All non-introgressed SNPs can be found by performing a right-join with a null key.
If only a representative sample is desired, it could be performed by selecting a random selection of introgressions from each population group, where each population group makes up a proportion of the sample equal to its proportion of the overall dataset.

### Issue 1.3
A Kruskal-Wallis test could be performed to test the assumption that the found counts for each individual group come from the same parental distribution. If any population groups were found to significantly diverge from the others, post-hoc pairwise Wilcox tests could be performed to determine which groups (if any) were divergent.

## New Progress Since Last Submission

## Project Organization

## Struggles and Requested Help
