## Prior Feedback
### Issue 1.1
The dataframe produced by alignSNPs.py has the following information:

- variantChromosome (Chromosome the variant is located on), variantPosition (Base pair location of the variant), variantAlt (The alternative base found for a given variant), variant_rsID (The unique identifier for a variant as assigned by NCBI), ensemblID (The gene ID for the gene a variant is located within), variantPIP (The probability that a particular variant is causal for a signal within a given credible set), NMATCH (If the variant was found to have introgressed from Neanderthals), DMATCH (If the variant was found to have introgressed from Denisovans), and POPGROUP (The population group in which a given introgression occured)

The log file for alignSNPs.py returns the number of matching SNPs found within a set search space in the following format:
- **Number of Matches Found** Alignments were found out of **Average number of introgressions of the searched-for match type across all population groups** introgressions across **Number of SNPs surveyed from selected eQTL finemapping dataset** SNPs.

Together, this information can be used in downstream analysis to filter for any selection of fine-mapped SNPs from any population group.

### Issue 1.2
All non-introgressed SNPs can be found by performing a outer merge of the eQTL-finemapping and SPrime dataframes and keeping values unique to the eQTL-finemapping.

If only a representative sample is desired, it could be selected by randomly selecting a desired sample size of rows from the non-introgressed dataframe, with each chomosome being represented in the sample according to its proportion within the population.

### Issue 1.3
A Kruskal-Wallis test could be performed to test the assumption that the found counts for each individual group come from the same parental distribution. If any population groups were found to significantly diverge from the others, post-hoc pairwise Wilcox tests could be performed to determine which groups (if any) were divergent.

### Issue 1.4

Yes, we will be performing a Mann-Whitney U Test to look at the median allele frequencies and/or SNP counts between the two groups. Depending on whether it is significant or not, it would be interesting to see whether Neanderthals and Denisovans have equally introgressed SNPs within the human population. 

## New Progress Since Last Submission

We started to make R code to plot Denisovan derived allele frequency versus number of SNPs, and will push to GitHub soon (weâ€™re still finalizing the script and working out some kinks). 

## Project Organization

Josh was in charge of aligning eQTL sets based on varying PIP thresholds with SNPs from the Sprime dataset based on whether they match with either Neanderthals or Denisovans.
Caroline and Fijare have been collaborating on R to plot out DAF vs. SNPs for both Neanderthals and Denisovans and will also be working on generating a combined graph (if possible) to compare the Neanderthal and Denisovan results.
Based on graphed data and statistical tests of identified SNPs, they will also conduct a literature search to find genes of interest that have introgressed into the human population.

## Struggles and Questions

Unsure whether the statistical test listed will be relevant and provide any meaningful analysis/comparison. Would love some feedback on this and whether you think there may be a better statistical test to compare the two groups?

What would be the best way to format a combined graph with both Neanderthal and Denisovan DAF vs. number of SNPs? 

